File: ocs_samples/library_samples/Python3/ocs_sample_library_preview/BaseClient.py
# BaseClient.py
#

import json

from .SdsError import SdsError

import requests
import time


class BaseClient(object):
    """Handles communication with Sds Service.  Internal Use"""

    def __init__(self, apiversion, tenant, url, clientId, clientSecret,
                 acceptVerbosity=False):
        self.__apiversion = apiversion
        self.__tenant = tenant
        self.__clientId = clientId
        self.__clientSecret = clientSecret
        self.__url = url  # if resource.endswith("/")  else resource + "/"

        self.__token = ""
        self.__expiration = 0
        self.__getToken()
        self.__acceptVerbosity = acceptVerbosity
        self.__requestTimeout = None

        self.__uri_API = url + '/api/' + apiversion

    @property
    def uri(self):
        """
        Gets the base url
        :return:
        """
        return self.__url

    @property
    def uri_API(self):
        """
        Returns the base URL plus api versioning information
        :return:
        """
        return self.__uri_API

    @property
    def api_version(self):
        """
        Returns just the base api versioning information
        :return:
        """
        return self.__apiversion

    @property
    def tenant(self):
        """
        Returns the tenant ID
        :return:
        """
        return self.__tenant

    @property
    def AcceptVerbosity(self):
        return self.__acceptVerbosity

    @AcceptVerbosity.setter
    def AcceptVerbosity(self, accept_verbosity):
        self.__acceptVerbosity = accept_verbosity

    @property
    def RequestTimeout(self):
        return self.__requestTimeout

    @RequestTimeout.setter
    def RequestTimeout(self, timeout):
        self.__requestTimeout = timeout

    def __getToken(self):
        """
        Gets the bearer token
        :return:
        """
        if ((self.__expiration - time.time()) > 5 * 60):
            return self.__token

        discoveryUrl = requests.get(
            self.__url + "/identity/.well-known/openid-configuration",
            headers={"Accept": "application/json"})

        if discoveryUrl.status_code < 200 or discoveryUrl.status_code >= 300:
            discoveryUrl.close()
            status = discoveryUrl.status_code
            reason = discoveryUrl.text
            raise SdsError(f"Failed to get access token endpoint "
                           f"from discovery URL: {status}:{reason}")

        tokenEndpoint = json.loads(discoveryUrl.content)["token_endpoint"]

        tokenInformation = requests.post(
            tokenEndpoint,
            data={"client_id": self.__clientId,
                  "client_secret": self.__clientSecret,
                  "grant_type": "client_credentials"})

        token = json.loads(tokenInformation.content)

        expiration = token.get("expires_in", None)
        if expiration is None:
            raise SdsError(f"Failed to get token, check client id/secret: {token['error']}")

        self.__expiration = float(expiration) + time.time()
        self.__token = token['access_token']
        return self.__token

    def sdsHeaders(self):
        """
        Gets the base headers needed for OCS call
        :return:
        """
        headers = {"Authorization": "Bearer %s" % self.__getToken(),
                   "Content-type": "application/json",
                   "Accept": "application/json"}
        if (self.__acceptVerbosity):
            headers['Accept-Verbosity'] = "verbose"
        if self.__requestTimeout is not None:
            headers['Request-Timeout'] = str(self.__requestTimeout)

        return headers

    def checkResponse(self, response, main_message):
        if response.status_code < 200 or response.status_code >= 300:
            status = response.status_code
            reason = response.text
            url = response.url
            opId = response.headers["Operation-Id"]
            error = f"  {status}:{reason}.  URL {url}  OperationId {opId}"
            response.close()

            message = main_message + error
            raise SdsError(message)

File: ocs_samples/library_samples/Java/ocs_sample_library_preview/src/main/java/com/github/osisoft/ocs_sample_library_preview/BaseClient.java
/** BaseClient.java
 * 
 */

package com.github.osisoft.ocs_sample_library_preview;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import java.io.*;
import java.net.*;
import java.nio.charset.StandardCharsets;
import java.util.Date;
import java.util.Properties;

/**
 * Base client that helps with interactions to OCS
 */
public class BaseClient {
    /**
     * accessible json converter
     */
    public Gson mGson = null;
    /**
     * use this to see the base url for going against OCS
     */
    public String baseUrl = null;
    /**
     * api version used for the calls
     */
    public String apiVersion = null;

    private String cachedAccessToken = null;
    private Date accessTokenExpiration = new Date(Long.MIN_VALUE);
    private long FIVE_SECONDS_IN_MILLISECONDS = 5000;


    //config parameters
    private String gclientId = "";
    private String gclientSecret = "";
    private String gresource = "";

    /**
     * Creates a baseclient.  Reading information from the configuration file at the program's running folder
     */
    public BaseClient() {
        gclientId = getConfiguration("clientId");
        gclientSecret = getConfiguration("clientSecret");
        gresource = getConfiguration("resource");
        gresource = gresource.endsWith("/") ? gresource :  gresource + "/";

        this.baseUrl = gresource;
        this.apiVersion = getConfiguration("apiVersion");
        this.mGson = new Gson();
    }
    
    /**
     * Creates a baseclient using the passed information rather than the configuration settings
     * 
     * @param apiVersion APIversion of OCS
     * @param clientId Client id to login with
     * @param clientSecret client secret to login with 
     * @param resource OCS url
     */
    public BaseClient(String apiVersion, String clientId, String clientSecret, String resource) {
        gclientId = clientId;
        gclientSecret = clientSecret;
        gresource = resource;
        gresource = gresource.endsWith("/") ? gresource :  gresource + "/";

        this.baseUrl = gresource;
        this.apiVersion = apiVersion;
        this.mGson = new Gson();
    }
   
    /**
     * Makes the connection to the url
     * @param url the url to connect to 
     * @param method the method to do, put, get, delete, etc...
     * @return
     */
    public HttpURLConnection getConnection(URL url, String method) {
        HttpURLConnection urlConnection = null;
        String token = AcquireAuthToken();

        try {
            urlConnection = (HttpURLConnection) url.openConnection();
            urlConnection.setRequestMethod(method);
            urlConnection.setRequestProperty("Accept", "*/*; q=1");
            urlConnection.setRequestProperty("Content-Type", "application/json");
            urlConnection.setRequestProperty("Authorization", "Bearer " + token);
            urlConnection.setUseCaches(false);
            urlConnection.setConnectTimeout(50000);
            urlConnection.setReadTimeout(50000);
            if ("POST".equals(method) || "PUT".equals(method) ||"DELETE".equals(method)) {
                urlConnection.setDoOutput(true);
            } else if (method == "GET") {
                //Do nothing
            }
        } catch (SocketTimeoutException e) {
            e.printStackTrace();
        } catch (ProtocolException e) {
            e.printStackTrace();
        } catch (IllegalStateException e) {
            e.printStackTrace();
        } catch (Exception e) {
            e.printStackTrace();
        }

        return urlConnection;
    }

    /**
     * Helper to get the bearer auth token
     * @return the token string
     */
    protected String AcquireAuthToken() {

        if (cachedAccessToken != null){
            long tokenExpirationTime = accessTokenExpiration.getTime(); // returns time in milliseconds.
            long currentTime = System.currentTimeMillis();
            long timeDifference = tokenExpirationTime - currentTime;

            if(timeDifference > FIVE_SECONDS_IN_MILLISECONDS)
                return cachedAccessToken;
        }

        // get new token 
        try {
            URL discoveryUrl = new URL(gresource + "identity/.well-known/openid-configuration");
            URLConnection request = discoveryUrl.openConnection();
            request.connect();
            JsonParser jp = new JsonParser(); 
            JsonObject rootObj = jp.parse(new InputStreamReader((InputStream) request.getContent(), StandardCharsets.UTF_8)).getAsJsonObject(); 
            String tokenUrl = rootObj.get("token_endpoint").getAsString(); 

            URL token = new URL(tokenUrl);
            HttpURLConnection tokenRequest = (HttpURLConnection) token.openConnection();
            tokenRequest.setRequestMethod("POST");
            tokenRequest.setRequestProperty("Accept", "application/json");
            tokenRequest.setDoOutput(true);
            tokenRequest.setDoInput(true);
            tokenRequest.setUseCaches(false);

            String postString = "client_id=" + URLEncoder.encode(gclientId, "UTF-8") 
                + "&client_secret=" + URLEncoder.encode(gclientSecret, "UTF-8") 
                + "&grant_type=client_credentials";
            byte[] postData = postString.getBytes("UTF-8");
            tokenRequest.setRequestProperty( "Content-Length", Integer.toString( postData.length));
            tokenRequest.getOutputStream().write(postData);

            InputStream in = new BufferedInputStream(tokenRequest.getInputStream());
            String result = org.apache.commons.io.IOUtils.toString(in, "UTF-8");
            in.close();

            jp = new JsonParser(); 
            JsonObject response = jp.parse(result).getAsJsonObject(); 
            cachedAccessToken = response.get("access_token").getAsString();
            Integer timeOut = response.get("expires_in").getAsInt();
            accessTokenExpiration = new Date(System.currentTimeMillis() + timeOut * 1000);
        } catch (MalformedURLException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }

        return cachedAccessToken;
    }

    /**
     * helper to get configuration information for the file
     * @param propertyId which property to retreive from the file
     * @return the value retreived
     */
    private String getConfiguration(String propertyId) {
        
         if(propertyId.equals("clientId") && !gclientId.isEmpty()){
            return gclientId;
        }

        if(propertyId.equals("clientSecret") && !gclientSecret.isEmpty()){
            return gclientSecret;
        }

        if(propertyId.equals("resource") && ! gresource.isEmpty()){
            return gresource;
        }

        String property = "";
        try(InputStream inputStream = new FileInputStream("config.properties")) {
            Properties props = new Properties();

            props.load(inputStream);
            property = props.getProperty(propertyId);
            inputStream.close();
        } catch (Exception e) {
            e.printStackTrace();
        }

        return property.trim();
    }

    /**
     * helper to check if the response code indicates success
     * @param responseCode code number
     * @return success
     */
    public boolean isSuccessResponseCode(int responseCode) {
        return responseCode >= 200 && responseCode < 300;
    }


}
